tutorial/tests/test_object_oriented_programming.py

```diff
@@ -1,93 +1,181 @@
+            self.scoops = []
+
+        def add_scoops(self, *new_scoops: "Scoop") -> None:
+            for one_scoop in new_scoops:
+                self.scoops.append(one_scoop)
+
+        def __str__(self):
+            return (
+                f"Ice cream bowl with {', '.join(s.flavor for s in self.scoops)} scoops"
+            )
+
+    bowl = Bowl()
+    scoops = [Scoop(flavor) for flavor in flavors]
+    bowl.add_scoops(*scoops)
+    return str(bowl)
+
+
+@pytest.mark.parametrize(
+    "flavors",
+    [
+        (("chocolate", "vanilla", "stracciatella")),
+    ],
+)
+def test_ice_cream_bowl(flavors, function_to_test) -> None:
+    assert function_to_test(flavors) == reference_ice_cream_bowl(flavors)
+
+
+#
+# Exercise 3: Ice cream shop
+#
+
+
+def reference_ice_cream_shop(flavors_1: list[str], flavors_2: list[str]) -> bool:
+    class Shop:
+        """A class representing an ice cream shop"""
+
+        def __init__(self, flavors):
+            self.flavors = flavors
+
+        def __str__(self):
+            return (
+                f"Ice cream shop selling flavors: {', '.join(f for f in self.flavors)}"
+            )
+
+        def __eq__(self, other):
+            if isinstance(other, Shop):
+                return len(self.flavors) == len(other.flavors)
+            return False
+
+        def __lt__(self, other):
+            if isinstance(other, Shop):
+                return len(self.flavors) < len(other.flavors)
+            return False
+
+        def __le__(self, other):
+            if isinstance(other, Shop):
+                return self < other or self == other
+            return False
+
+    shop_1 = Shop(flavors_1)
+    shop_2 = Shop(flavors_2)
+    return shop_1 <= shop_2
+
+
+@pytest.mark.parametrize(
+    "flavors_1, flavors_2",
+    [
+        (["chocolate", "vanilla", "stracciatella"], ["caramel", "strawberry", "mango"]),
+        (["vanilla", "stracciatella"], ["chocolate", "vanilla", "mango"]),
+        (["vanilla", "mango"], ["chocolate"]),
+    ],
+)
+def test_ice_cream_shop(flavors_1, flavors_2, function_to_test) -> None:
+    assert function_to_test(flavors_1, flavors_2) == reference_ice_cream_shop(
+        flavors_1, flavors_2
+    )
+
+
+#
+# Exercise 4: Intcode computer
+#
+
+
+def read_data(name: str, data_dir: str = "data") -> pathlib.Path:
+    """Read input data"""
+    return (pathlib.Path(__file__).parent / f"{data_dir}/{name}").resolve()
+
+
+def prepare_params() -> list[str]:
+    intcodes = ["1,0,0,0,99", "2,3,0,3,99", "1,1,1,4,99,5,6,0,99"]
+    intcodes += [read_data(f"intcode_{i}.txt").read_text() for i in (1, 2)]
+    return intcodes
+
+
+def reference_intcode_computer(intcode: str) -> int:
+    class Computer:
+        """An Intcode computer class"""
+
+        def __init__(self, program: str):
+            self.program = [int(c.strip()) for c in program.split(",")]
+            self._backup = self.program[:]
+
+        def reset(self):
+            self.program = self._backup[:]
+

```

Comment:
These lines are useless. We should remove them from the solution to avoid confusion.

```suggestion
```